### ISSUE TITLE
Dealing "Cannot render cell; GraphQL success but `data` is null" errors

### ISSUE THREAD

**Issue by jangxyz** [redwoodjs/redwood#2473]
From the discussion: [Dealing with “GraphQL success but `data` is null” in storybook](https://community.redwoodjs.com/t/dealing-with-graphql-success-but-data-is-null-in-storybook/2098)

I get to face this error message when mock data is somehow not aligned with the query:

    Cannot render cell; GraphQL success but `data` is null

![image](https://user-images.githubusercontent.com/52015/117477121-8ff4b180-af98-11eb-9b28-7c284b0107b6.png)

This is user error, but I think it will be better if there is a way to reveal what is going on, as it is quite difficult to figure out the cause only with the message.

**Comment by peterp** [comment]
I've actually experienced this too,  and I wrote the code. I think @dac09 has a valid point that this can be fixed with types, we're very close to getting our Cell's QUERY types.

But, I'm wondering if we shouldn't double down and improve the errors at run-time? I can imagine a scenario where you change the query, but don't update the mock, and then a few weeks later attempting to run Storybook and you're unable to determine what the error actually means.

**Comment by dac09** [comment]
If we can find the specific type of error - a better explained error would sure help on top of the types!

**Comment by noire-munich** [comment]
I just got this one - upgraded to v0.34 this morning from v0.33.
Also: storybook is not involved in my user case.
Quick dirty fix was to change my Success component's signature so it receives a data object and not spread properties, then I test it before use in my render method.

Don't like it though.

**Comment by peterp** [comment]
@noire-munich Can you show us what you changed?

**Comment by noire-munich** [comment]
@peterp ah! I thought I could rewind but the error doesn't occur anymore on that component - it does on another one and I haven't been able to fix it yet.

**Comment by noire-munich** [comment]
@peterp @thedavidprice any chance this could be prioritized ? It's blocking further developments on SportOffice.
Should I find anything useful about it I'll report here.

**Comment by peterp** [comment]
@noire-munich I'm trying to figure out how to reproduce it, is it because the props passed into success are undefined?

**Comment by noire-munich** [comment]
@peterp They are not supposed to be undefined, that's the payload:

```json
{"data":{"structure":{"id":1,"createdAt":"2021-06-23T14:14:45.078Z","updatedAt":null,"slug":"new-account-verbyle","label":"Verbyle","currency":"EUR","planning":null,"time":null,"tunnel":{"id":8,"__typename":"Tunnel"},"__typename":"Structure"}}}
```

That's my Query:
```js
export const QUERY = gql`
  query EDIT_STRUCTURE($id: Int!) {
    structure: structure(id: $id) {
      id
      createdAt
      updatedAt
      slug
      label
      currency
      planning {
        startsAt
        endsAt
      }
      time {
        startsAt
        endsAt
      }
      tunnel {
        id
      }
    }
  }
`
```

And that's the Success declaration:
```js

export const Success = ({ id, structure }) => {
  const { t } = useTranslation()
# [...] code below would be irrelevant, unless specifically required.
```

**Comment by jangxyz** [comment]
If it helps, I got the feeling it happens when I add some of the fields in the query in some part and didn't update other parts (like the mock for the story). I figured that was the reason apollo was complaining for -- graphql query has succeeded, but insufficient data.

**Comment by peterp** [comment]
@noire-munich nope - I absolutely agree that they shouldn't be undefined, and this is definitely a bug - it just wasn't clear to me what the root cause is. 

My question was trying to establish that.

@jtoar can we figure this out?

**Comment by jtoar** [comment]
@peterp on it now, @noire-munich I'll try to reproduce with your example and ask for more info if needed!

**Comment by jtoar** [comment]
@noire-munich I haven't been able to reproduce this outside of storybook yet. For the error to be thrown:
```
'Cannot render cell: GraphQL success but `data` is null'
```
data has to be `null` or `undefined`. Data's `undefined` if Apollo can't get the query in it's entirety. (The default [errorPolicy](https://www.apollographql.com/docs/react/data/error-handling/#graphql-error-policies) is `'none'`).

One thing you could do to try to debug this is to set the errorPolicy in beforeQuery—maybe it'll show us what field it's having a hard time retrieving:

```js
export const beforeQuery = (props) => {
  return {
    variables: props,
    fetchPolicy: 'cache-and-network',
    notifyOnNetworkStatusChange: true,
    errorPolicy: 'all'
  }
}
```

If `'all'` doesn't work, you can also try `'ignore'`.

I'll keep trying things in the meantime!

**Comment by peterp** [comment]
@jtoar I'm wondering why it would hit success if data is null? Doesn't that already indicate some sort of error?

**Comment by peterp** [comment]
@jtoar Do you think this might have something to do with the `updating` change that we made? That in a subsequent fetch "structure" is undefined?

**Comment by jtoar** [comment]
> I guess null can be a valid response if it's something that's an optional return

Right! Moreover, if data is `null`, the Cell should render Empty based on the logic here:
https://github.com/redwoodjs/redwood/blob/162daf5acb3204df9049c9605cadacfbe6d5c2af/packages/web/src/components/createCell.tsx#L91-L93

I think I've seen Empty being rendered by data being an empty array a lot more than I've seen it being rendered by data being null. 

For Storybook, when you mismatch the fields on your Cell's `QUERY` (@jangxyz's example), we get the curious case that both `error` and `data` are undefined:

![image](https://user-images.githubusercontent.com/32992335/123158865-28e46c80-d421-11eb-95b5-ef036244c0db.png)

I feel like there should be an error here. 

---

> Do you think this might have something to do with the updating change that we made? That in a subsequent fetch "structure" is undefined?

Definitely possible, but if data really is undefined on a subsequent fetch, Apollo Client should be sending us back an error so we can render Failure instead. I'm confused why that's not the case in outside Storybook. I'll try intentionally breaking the resolver somehow

**Comment by jtoar** [comment]
Here's a walk through of a stress test. I also tried returning `null` instead of just omitting the email field, and Empty is rendered appropriately

https://user-images.githubusercontent.com/32992335/123169907-ba0e1000-d42e-11eb-91a5-50f7837e4951.mp4

**Comment by noire-munich** [comment]
@jtoar I tried with both `all` and `ignore` but it didn't change anything and didn't help gather more information on the issue at hand.

Also, it appears the error happens deeper in the page, the EditCell calls several cells, one of which being the LevelsCell, which has that query:
```js
export const QUERY = gql`
  query STRUCTURE_DISCIPLINE_LEVELS($structureId: Int!) {
    structure(id: $structureId) {
      disciplines {
        id
        label
        levels {
          label
          id
        }
      }
    }
  }
`
```

We use the same endpoint as in my initial message but we filter on different properties.
Here's the service:

```ts
export const structure = ({ id }: StructureWhereUniqueInput) => {
  logger.info(`Structures.structure`)

  return db.structure
    .findMany({
      take: 1,
      where: { id, isRemoved: false },
      include: {
        periods: {
          select: { id: true, season: true, startsAt: true, endsAt: true },
        },
        disciplines: {
          select: {
            id: true,
            label: true,
            levels: { select: { id: true, label: true } },
          },
        },
        time: true,
        planning: { select: { startsAt: true, endsAt: true } },
        tunnel: { select: { id: true } },
      },
    })
    .then(([structure]) => structure)
}
```

And the payload:
```json
{"data":{"structure":{"disciplines":[],"__typename":"Structure"}}}
```

On other cells I have no issue displaying an Empty or a Failure component.
I've tried making each entity optional/required in the related `sdl` but it didn't do any good.

**Comment by jtoar** [comment]
@noire-munich dang sorry that didn't work—and thanks for the info, I'll try nesting the Cells and nesting the payload. Superficially, the nested payload reminds me of this issue: https://github.com/redwoodjs/redwood/issues/2005. That wasn't causing anything to error, just wanted you to know that there's a discussion around routing a nested payload to Empty. 

I'll try reproducing this again!

**Comment by noire-munich** [comment]
Thanks @jtoar , just had a look at #2005 and indeed it's good to know others are considering such a thing ;).

Out of curiosity I checked the query call in `localhost:8911/graphql`, same arguments, I got the same payload in the playground and no error whatsoever - if it was apollo related I'd have expected something here.

But! There could be another thing to consider: in this context, I'm using a query that's already been called successfully yet with different fields, maybe that could lead to empty `data`? I'm unsure how to test though, I've seen this: https://www.apollographql.com/docs/react/caching/cache-configuration/#resultcaching, it could be worth a try to switch it to false but I'm not sure how I should modify `App.ts` `<RedwoodApolloProvider>` to do this, if relevant at all.

**Comment by noire-munich** [comment]
Following a debugging session with @jtoar and for the sake of public documentation, here's a temporary fix until a proper patch is provided. Please note that it's relevant if the bug happens in your app, not within Storybook.

```js
export const beforeQuery = (props) => {
  return { variables: props, fetchPolicy: 'no-cache' }
}
```

To put in the faulty cell with a warning.

Also, if anyone's working on this, it could be interesting to share if the ApolloClient's cache has actual data in it or not - mine hasn't.

**Comment by thedavidprice** [comment]
@jtoar are there any actionable next-steps for this one?

**Comment by thedavidprice** [comment]
Possibly related to #2981

**Comment by cjreimer** [comment]
FYI, we hit this bug as well with 0.34.1, not related to storybook.  Changing the fetchPolicy also works as a temporary fix of the issue for us.  However, was looking pretty tough to debug before I noted the workaround in this thread.

Our query looks like this:

```
  query FindProjectSettingQuery($id: Int!) {
    projectSetting: projectSetting(id: $id) {
      id
      project {
        name
        description
        owner
      }
      disclaimer {
        id
        body
      }
      disclaimerId
      startDate
      plannedEndDate
    }
  }

```
In our case, if we loaded the query cold, that is shortly after app initialization, we got the error.  But if we ran a different query to load the disclaimers, and then ran this query, it would work.  Seems like a strange cache issue.  Note that the query returns the following to `Success `when it is working:

```
disclaimer: {id: 1, body: "This review is only for general conformance ....", __typename: "SubmittalReviewDisclaimer"}
disclaimerId: 1
id: 1
plannedEndDate: "2022-06-01T22:00:00.000Z"
project: {name: "WTP...", description: "Upgrades to the....", owner: "R.M. of ....", __typename: "Project"}
startDate: "2020-02-01T23:00:00.000Z"
__typename: "ProjectSetting"
__proto__: Object
```

**Comment by noire-munich** [comment]
@thedavidprice @jtoar any updates on this one?

**Comment by dthyresson** [comment]
@thedavidprice and @jtoar This issue came up today in Discord from @viperfx.

Can we bump priority?

**Comment by KrisCoulson** [comment]
@thedavidprice  @jtoar  @dthyresson I actually just hit this as well for the first time today.

**Comment by thedavidprice** [comment]
Bumping priority on this to look at for next release sprint (v0.37)

**Comment by alecswan** [comment]
I am new to redwood and ran into a similar error while following [Cells section in RW tutorial](https://learn.redwoodjs.com/docs/tutorial/cells) and subsequently adding the following test. I tried changing the format of the mockGraphQLQuery response without any luck. It's also possible that testing of `render(<BlogPostsCell />)` is not supported.
 ```
 it('renders Title from fake GraphQL query successfully', async () => {
    // see https://redwoodjs.com/docs/testing#mockgraphqlquery
    mockGraphQLQuery('BlogPostsQuery', () => {
      return {
        data: {
          posts: [{ id: 1, title: 'My Test 1' }, { id: 43, title: 'My Test 2' }]
        }
      }
    })

    render(<BlogPostsCell />)

    expect(await screen.findByText('My Test 1')).toBeInTheDocument()
  })
```
Error stack trace is shown below:
```
BlogPostsCell
    ✓ renders Loading successfully (107 ms)
    ✓ renders Empty successfully (4 ms)
    ✓ renders Failure successfully (5 ms)
    ✓ renders Success successfully (7 ms)
    ✓ renders mock Title successfully (11 ms)
    ✕ renders Title from fake GraphQL query successfully (87 ms)

  ● BlogPostsCell › renders Title from fake GraphQL query successfully

    Cannot render cell: GraphQL success but `data` is null

      at ../node_modules/@redwoodjs/web/dist/components/createCell.js:140:15
      at Query (../node_modules/@redwoodjs/web/dist/components/createCell.js:29:19)
      at renderWithHooks (../node_modules/react-dom/cjs/react-dom.development.js:14985:18)
      at updateFunctionComponent (../node_modules/react-dom/cjs/react-dom.development.js:17356:20)
      at beginWork (../node_modules/react-dom/cjs/react-dom.development.js:19063:16)
      at HTMLUnknownElement.callCallback (../node_modules/react-dom/cjs/react-dom.development.js:3945:14)
      at HTMLUnknownElement.callTheUserObjectsOperation (../node_modules/jsdom/lib/jsdom/living/generated/EventListener.js:26:30)
      at innerInvokeEventListeners (../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:338:25)
      at invokeEventListeners (../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:274:3)
      at HTMLUnknownElementImpl._dispatch (../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:221:9)
      at HTMLUnknownElementImpl.dispatchEvent (../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:94:17)
      at HTMLUnknownElement.dispatchEvent (../node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:231:34)
      at Object.invokeGuardedCallbackDev (../node_modules/react-dom/cjs/react-dom.development.js:3994:16)
      at invokeGuardedCallback (../node_modules/react-dom/cjs/react-dom.development.js:4056:31)
      at beginWork$1 (../node_modules/react-dom/cjs/react-dom.development.js:23964:7)
      at performUnitOfWork (../node_modules/react-dom/cjs/react-dom.development.js:22779:12)
      at workLoopSync (../node_modules/react-dom/cjs/react-dom.development.js:22707:5)
      at renderRootSync (../node_modules/react-dom/cjs/react-dom.development.js:22670:7)
      at performSyncWorkOnRoot (../node_modules/react-dom/cjs/react-dom.development.js:22293:18)
      at ../node_modules/react-dom/cjs/react-dom.development.js:11327:26
      at unstable_runWithPriority (../node_modules/scheduler/cjs/scheduler.development.js:468:12)
      at runWithPriority$1 (../node_modules/react-dom/cjs/react-dom.development.js:11276:10)
      at flushSyncCallbackQueueImpl (../node_modules/react-dom/cjs/react-dom.development.js:11322:9)
      at flushSyncCallbackQueue (../node_modules/react-dom/cjs/react-dom.development.js:11309:3)
      at scheduleUpdateOnFiber (../node_modules/react-dom/cjs/react-dom.development.js:21893:9)
      at dispatchAction (../node_modules/react-dom/cjs/react-dom.development.js:16139:5)
      at ../node_modules/@apollo/client/react/hooks/utils/useBaseQuery.js:17:85

  console.error
    Error: Uncaught [Error: Cannot render cell: GraphQL success but `data` is null]
```

**Comment by dthyresson** [comment]
Hi @jangxyz I have not tried this yet, but because the `BlogPostsQuery` expects the data do contains `blogPosts` and not `posts`, that may be the reason for the `but `data` is null` error.

```js
// web/src/components/BlogPostsCell/BlogPostsCell.js

export const QUERY = gql`
  query BlogPostsQuery {
    blogPosts {
      id
    }
  }
`

export const Success = ({ blogPosts }) => {
  return JSON.stringify(blogPosts)
}```

Could you try:

```js
   mockGraphQLQuery('BlogPostsQuery', () => {
     return {
       data: {
         blogPosts: [{ id: 1, title: 'My Test 1' }, { id: 43, title: 'My Test 2' }]
       }
     }
   })```

and let me know if that works instead?

**Comment by alecswan** [comment]
Hi @dthyresson, your suggestion did not work but its understandable because [Cells section in RW tutorial](https://learn.redwoodjs.com/docs/tutorial/cells) recommends renaming blogPosts to posts inside BlogPostsQuery half-way through that section, which I did. I should have included the source code in my previous comment to avoid any confusion. Please find it below.

```
export const QUERY = gql`
  query BlogPostsQuery {
    posts {
      id
      title
      body
      createdAt
    }
  }
`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ posts }) => {
  return posts.map((post) => (
    <article key={post.id}>
      <header>
        <h2>{post.title}</h2>
      </header>
      <p>{post.body}</p>
      <div>Posted at: {post.createdAt}</div>
    </article>
  ))
}
```

**Comment by KrisCoulson** [comment]
I actually just hit this again and it took me quite a long time to figure out again. I believe the answer that @pi0neerpat recommends would have fixed it but interesting detail the way I triggered this was in a react-select and I wasn't even using a cell... I guess maybe it is apollo cache needs the exact same grapqhl signature as something else that was using maybe the same key I can def elaborate more if we need more details

**Comment by jtoar** [comment]
@KrisCoulson I've been looking at this again with @callingmedic911. One thing that would help us a lot is a reproducible example. Is there any way you could extract and/or share the part of your app that ran into this error? I'm happy to help you with that process!

Another bit of information that would help us is your Apollo Client version. Are you on the latest? (3.4.16, which is what v0.38 uses)

**Comment by mojombo** [comment]
This might be a good chance to try out Replay to capture the problem and send it to us! https://www.replay.io/

**Comment by thedavidprice** [comment]
@jtoar sound like this has been resolved via Apollo Client 3.4, correct?

**Comment by jtoar** [comment]
@thedavidprice I believe so; it's still probably worth improving the error message and setting the verbosity of the Apollo logs to debug in case this happens again. We could close this out and either @callingmedic911 or I could follow up with a PR implementing those two things

**Comment by callingmedic911** [comment]
**Summary:** (if we need to come back here, hopefully never 😆)

For the actual response (not from the storybook), it's most likely a bug that is fixed upstream and therefore in RedwoodJS v0.38.0: https://github.com/apollographql/apollo-client/issues/8370. Relevant issue:  https://github.com/apollographql/apollo-client/issues/8063

Other possible cases this may happen is if there’s a difference in expected query fields and the `response`. `response` could be from:
- Mocked (For example from Storybook): Probably some missing field in mock data?
- Cached: If the `fetchPolicy` is set to cache-only and cache is incomplete/missing, then this could happen.
- Actual response from the server: Highly unlikely but if the server fails to produce a GraphQL spec valid response.
